[DataGridPremium] Support sort-dependent aggregation and improve performance

lauri865 opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

<!-- You can use `## Changelog` to create a description for this change in the next release. -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Fixes #18342 (the first part of it)

Adds a new `applySorting?: boolean` property to `GridAggregationFunction`

The only thing that probably needs additional polish is this:
```typescript
useGridEvent(apiRef, 'filteredRowsSet', applyAggregation);
useGridEvent(apiRef, 'sortedRowsSet', applyAggregation);
```

These can in some instances (on mount) be triggered sequentially if I remember correctly (correct me if I'm wrong), so might be some room for optimisation.

Changes:
* Optimized columns updates (-800ms with a large amount of columns)
* Optimized column group creation (50-100x improvement under large number of columns)
* Improved aggregation performance (from 6000ms+ -> 275ms in my test case)
* Removed `pivotModeChange` control state, because `pivotMode` is not strictly controlled, and in an uncontrolled mode, it delays updates by 1 render
* Fixed a case whereby rows are improperly restored (props.rows takes overrides any subsequent changes made via the apiRef)
* Created windowed and chunked aggregation logic, which calculates the current viewport instantly, and the rest in chunks to avoid locking the main thread. Chunksize logic could be improved further; one idea is to continue with the chunks until we hit X amount of ms, or recalculate chunkSize based on how long the previous chunk took.
* Deduplicated calls to `applyAggregation` (currently it was called many times in a row)
* Support for `applySorting` in aggregation functions
* Expose `field` to `getCellValue` in aggregation functions to make custom logic possible, especially in the context of pivoting where fields are automatically generated
* Fixed: column groups based on numbers were not sorted correctly, as all headernames were coerced to strings
* Fixed: pivoting doesn't initialize when rows are updated with `setRows` while `isLoading` internal state is still true. Common pattern would be to set rows within `useLayoutEffect` when `isLoading` is changing (to avoid flickering), and that can't be caught by the current logic since an upperlevel `useLayoutEffect` will always run before the update of the internal state. Deferred initializing `nonPivotDataRef` to the first row update (works with all of `setRows`,  `updateRows` and `props.rows`) instead of listening to the isLoading change of internal state.

Performance updates; before:

https://github.com/user-attachments/assets/34bd24b8-7f26-467e-83a9-9478e320adfc

After:

https://github.com/user-attachments/assets/d6985bdb-b078-4f84-99b3-913cbd85b9bd

lauri865 commented:
Enabling pivot seems to trigger 5 instances of "filteredRowsSet" (each of them retriggers aggregations). And the same for "sortedRowsSet for whatever reason @cherniavskii 

Adding `applySorting` to the mix causes a visible slowdown.

Just press the toggle button and check console:
https://stackblitz.com/edit/g4dsphrl-sxhvdgqz?file=src%2FDemo.tsx,package.json

lauri865 commented:
I tried deduplicating some more expensive calls, but probably broke some tests (not sure if for functionality, but at least it will complain about `act()`.

One thing I still don't understand though – when stopping pivoting, columns get rehydrated 6 times:
6x `MUI X: useGridColumns - Columns pipe processing have changed, regenerating the columns`

lauri865 commented:
Did a bunch of optimisations in aggregation calculations. For a a large pivot table (10k rows, pivoted by date (value) and commodity in the dataset, summing "total in usd") – computation time went down from 6000ms+ to 375ms. A big culprit was nested function calls to access row values, which we can skip for pivoting + doing another full recursion over all rows/columns. isNumber custom function wrapper also added quite significant overhead for summing.

Can be still polished a bit I guess.

lauri865 commented:
From 375ms -> 260ms in my test case now that I stoped storing undefined values. I guess it's fair to assume that undefined has no meaning in aggreations, nulls might (although even that is slightly questionable, and could be optionally toggled). And there's tons of undefined values in pivoting at least.

lauri865 commented:
Fixed the last issue as well with cascading `hydrateColumns` updates by re-registering the `pipeApplier` when `pivotMode` changes. So the slow restore after pivoting in the after video should also hopefully be fixed now.

lauri865 commented:
This took about 800ms when toggling a pivot with a large list of columns, now down to <1ms:
https://github.com/mui/mui-x/blob/8d434ffcb9d091293a682d79be94a81a9aa66812/packages/x-data-grid/src/hooks/features/columns/gridColumnsUtils.ts#L347-L353

There's still some room to optimise – 168ms just for columns + aggregation + row grouping – easy to lock up the main thread still:
* useGridColumns createColumnsState: 30.65087890625 ms
* useGridColumns setGridColumnsState: 138.465087890625 ms

lauri865 commented:
Toggling off pivoting doesn't lock up the UI anymore:

https://github.com/user-attachments/assets/e31021ba-b56f-4501-abe3-4ea9c173fec3

lauri865 commented:
Implemented viewport aware, chunked aggregations to avoid locking. Switch animation runs buttery smooth even with 10k rows. With 100k rows it's still usable, but there's a slight lockup, depending on the dimensionality, but the bottleneck is not in aggregation anymore.

lauri865 commented:
I'm not sure how to satisfy the tests without a lot of work. The quemicrotask and settimeout makes the test fail. Don't feel particularly like (=dont have the time for) updating all the  aggregation and pivoting tests with mock timers.

cherniavskii commented on packages/x-data-grid-premium/src/hooks/features/aggregation/createAggregationLookup.ts:
```diff
@@ -62,39 +87,46 @@ const getGroupAggregatedValue = (
-        aggregatedValues[j].values.push(aggregationFunction.getCellValue({ row }));
+        value = aggregationFunction.getCellValue({ field, row });
+      } else if (isPivotActive) {
+        // Since we know that pivoted fields are flat, we can use the row directly, and save lots of processing time
```
That's a good assumption since all `valueGetter`s are being called when building the pivoted set of rows 👍🏻

cherniavskii commented on packages/x-data-grid-generator/src/hooks/serverUtils.ts:
```diff
@@ -312,11 +312,14 @@ const applyAggregation = (
+        field,
+        groupId,
+      },
+      null as any, // we can't use apiRef here, and as long as we don't use advanced aggregation functions that depend on it, we're clear
```
Yeah, that's fine.

> I also exposed api in aggregationFunction.apply, as I've felt many times that I could build more flexible and performant aggregations if only I had more context to go by.

BTW, do you have some specific examples you can share?

lauri865 replied:
Sure, let's say you want to apply a weighted average calculation, and make the weight column configurable by the end-user. How would you achieve this?

cherniavskii replied:
I would use a custom aggregation function with `getCellValue` and dynamically access the weights depending on the selected weight column. Something like this: https://stackblitz.com/edit/ucpxtjbq?file=src%2FDemo.tsx
If necessary, you could also access `apiRef` in the custom aggregation function.
WDYT?

cherniavskii reviewed (commented):
Thanks for the PR!
It's a lot to review, so I'll give it a few more passes. But overall, looks solid!

lauri865 commented:
Fixed another issue that was bugging us that prevented pivoting from initializing when setting rows with `apiRef.current.setRows` while `isLoading` is changing from true -> false.

Also added a path to re-run state initializers to avoid cascading and conflicting states (e.g. changing `rowGroupingModel` ahead of setting rows would recalculate row groups on the old data, and the same for aggregation). That made brough a 100k row pivot to quite acceptable performance levels.

Note though: controlled mode is still quite unoptimized. It calculates the pivoting state, throws it away (ignores controlled state being set), and recalculates it once the upper-level state/prop updates (cc @cherniavskii).

lauri865 commented on packages/x-data-grid-premium/src/hooks/features/pivoting/useGridPivoting.ts:
I think the first value of column's `sortingOrder` would be great.

Controlled prop: good point, however, since we treat pivoting model as persistent state (users can use it to explore data, navigate away, come back and continue), we'd rather not add a controlled prop for it (due to past scar tissue with performance and ease of restoring).

romgrk commented on packages/x-data-grid/src/hooks/utils/useRunOncePerLoop.ts:
```diff
@@ -0,0 +1,41 @@
+        callback(...args);
+      };
+
+      if (nextFrame) {
+        if (typeof requestAnimationFrame === 'function') {
+          requestAnimationFrame(runner);
+        }
+        return;
+      }
+
+      if (isJSDOM) {
+        runner();
+      } else if (typeof queueMicrotask === 'function') {
+        queueMicrotask(runner);
+      } else {
+        Promise.resolve().then(runner);
+      }
```
The name of the hook doesn't exactly match its implementation. The microtask queue is run until it's empty, so if the scheduled callback somehow schedules itself again, it could run multiple times in a single tick of the event loop. I don't see the animation frame option being used. If that was removed, this hook could be renamed to something like `useScheduleAsMicrotask`. If what we need is really to run at most once per tick, then it should use a task, not a microtask.

romgrk replied:
lint: `queueMicrotask` seems to be available in all our supported browsers, we can use it without guard.

lauri865 replied:
> The name of the hook doesn't exactly match its implementation. The microtask queue is run until it's empty, so if the scheduled callback somehow schedules itself again, it could run multiple times in a single tick of the event loop. I don't see the animation frame option being used. If that was removed, this hook could be renamed to something like `useScheduleAsMicrotask`. If what we need is really to run at most once per tick, then it should use a task, not a microtask.

Can you bring an example of how this is possible within the context of React? QueMicrotsk should put it at the end of the call stack, the only way it can schedule itself again is a straight up recursion that doesn't set any state (a state change would trigger a new render, and the microtask would run before that). Which is a) unlikely b) fine, since we want to avoid expensive calculations in the same frame that compute new state.

romgrk replied:
> Could be the behaviour of React in the past, but with batched state updates, that can't be the case, can it? How would batching work if a state update would trigger a back-to-back render?

If a state update is triggered within an effect, React could determine that it still has time to run and trigger a new render phase. Besides, batching doesn't necessarily imply using microtasks, because React is at the root of the call stack it can batch updates done within the full render cycle without resorting to microtasks.

Anyway, if we can't find a better name I don't want to cause too much bikeshedding, the current name is acceptable.

lauri865 commented:
Also implemented `exportState` with pivoting (we really needed it, since otherwise restored state is messed up).

cherniavskii reviewed (approved):
The changes look good to me, I'll look into failing tests!
Thanks again, Lauri!

cherniavskii commented on packages/x-data-grid-pro/src/tests/detailPanel.DataGridPro.test.tsx:
```diff
@@ -325,8 +328,9 @@ describe('<DataGridPro /> - Detail panel', () => {
+    // + 2x during state initialization (StrictMode)
     // + 2x when sortedRowsSet is fired
-    const expectedCallCount = reactMajor >= 19 ? 4 : 8;
+    const expectedCallCount = reactMajor >= 19 ? 6 : 12;
```
I'm not exactly sure why this needed to be updated...

lauri865 commented on packages/x-data-grid/src/hooks/core/pipeProcessing/useGridPipeProcessing.ts:
```diff
@@ -65,6 +69,8 @@ export const useGridPipeProcessing = (apiRef: RefObject<GridPrivateApiCommon>) =
     isRunning.current = false;
   }, []);
 
+  const runAppliersDeferred = useRunOncePerLoop(runAppliers, false, getGroupScopeKey);
```
Is there a case in which this could introduce flickering or other unwanted behaviour – e.g. some changes get flushed to DOM based on updated state, but pipe processing takes place after that (e.g. state changes -> DOM updates based on the change state -> columns get rehydrated only on the next render). Essentially the equivalent of setting state in `useEffect` vs. `useLayoutEffect`.

In the case of aggregation it's fine (especially with batched processing now), but broadly applying it to all pipe processors might make be a bit more questionable / tricky.

Alternative approach for this case would be to mark a group as pending for updates, and add a hook as the last entry in `useDatagridPremiumComponent` that clears the queue / triggers `runAppliers`.

lauri865 replied:
Or actually, the deduplication can be even easier for pipe processing – just accept the first call to `runPipeApplier` per render and set a flag to prevent further calls, which gets cleared in a `useEffect`.

Microtasks are needed only in relation to events, as there's a timing issue. Pipe processors depend on state / props, which are share across all hooks in the same render pass / component.

cherniavskii commented:
@lauri865 Thank you for making it possible, we appreciate it!
